src/services/hsm.service.js

```diff
@@ -6,16 +6,210 @@
+  return getBip32Keyname(keyname, 3);
+}
+
+function getChangeKeyname(keyname) {
+  return getBip32Keyname(keyname, 4);
+}
+
+/**
+ * @typedef {Object} SignatureData
+ * @property {Buffer} signature
+ * @property {Buffer} pubkey
+ * @property {number} addressIndex
+ * @property {number} inputIndex
+ */
+
+class HsmSession {
+  /**
+   * @param {hsm.interfaces.Hsm} conn
+   * @param {string} rootKeyname
+   */
+  constructor(conn, rootKeyname) {
+    this.conn = conn;
+    this.rootKey = rootKeyname;
+    /**
+     * @type {Record<number, string>}
+     */
+    this.addressKeys = {};
+  }
+
+  /**
+   * Get the adddress index keyname from the address index
+   * @param {index} index - bip32 Address index
+   * @returns {Promise<string>}
+   */
+  async getAddressKeyName(index) {
+    if (index in this.addressKeys) {
+      return this.addressKeys[index];
+    }
+
+    const addressKeyName = getBip32Keyname(this.rootKey, 5, { index });
+    const changeKeyName = getChangeKeyname(this.rootKey);
+    try {
+      await this.conn.blockchain.createBip32ChildKeyDerivation(
+        hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+        index,
+        false,
+        true,
+        changeKeyName,
+        addressKeyName,
+      );
+    } catch (err) /* istanbul ignore next */ {
+      // After an error, the HSM lib has a runtime issue that will fail the next call
+      // To avoid this we can wait 500ms before continuing.
+      // This may be fixed in the near future.
+      await delay(500);
+      // (CODE: 5022 | NAME: ERR_OBJ_ALREADY_EXISTS)
+      // This error code means the address key was already derived
+      if (err.errorCode === hsm.constants.ERR_OBJ_ALREADY_EXISTS) {
+        // address key already exists, just return it
+        this.addressKeys[index] = addressKeyName;
+        return addressKeyName;
+      }
+      console.error(err);
+      throw err;
+    }
+
+    this.addressKeys[index] = addressKeyName;
+    return addressKeyName;
+  }
+
+  /**
+   * Sign the given data with the given HSM key
+   * @param {HsmSession} hsmSession
+   * @param {Buffer} dataToSignHash
+   * @param {number} addressIndex
+   * @returns {Promise<Buffer>} DER encoded signature
+   */
+  async signData(dataToSignHash, addressIndex) {
+    // Derive the key to the desired address index
+    const htrKeyName = await this.getAddressKeyName(addressIndex);
+
+    // This api receives the pre-hashed data to sign
+    // We request a non-deterministic ECDSA signature
+    // and it returns the DER encoded signature Buffer
+    const signature = await this.conn.blockchain.sign(
+      hsm.enums.BLOCKCHAIN_SIG_TYPE.SIG_DER_ECDSA,
+      hsm.enums.BLOCKCHAIN_HASH_MODE.SHA256,
+      dataToSignHash,
+      htrKeyName,
+    );
+
+    // Dinamo SDK returns v + DER, where v is a parity byte
+    // https://manual.dinamonetworks.io/nodejs/enums/hsm.enums.BLOCKCHAIN_SIG_TYPE.html#SIG_DER_ECDSA
+    // We can safely ignore the parity byte
+    if (signature[0] !== 0x30) {
```

Comment:
I read an explanation about this value in another commend you replied to Abadesso, can you add it in the code?

> The 0x30 in DER encoding is a tag byte that marks the subsequent data as a sequence (or array). Since a DER signature is the [r,s] array it should always start with 0x30, but dinamo adds a parity byte.

Reply:
Done